### PR TITLE
Fix error resposes for `from` search prefix

### PR DESCRIPTION
**Pull request by aescling**
Addresses #17941.

Using unsupported prefixes now reports a 422; searching for posts from an account the instance is not aware of reports a 404. I think the UI for this on the front end is only barely more helpful at actually communicating any problems to the user, but at least the semantics are reasonable.

In addition, this commit addresses two minor points. More importantly, `from:username@domain` succeeds when `domain` is the local domain. Less importantly, `from:@username(@domain)?` is now valid syntax.

For the record, a nealy identical version of this commit in https://github.com/kibicat/mastodon/commit/945c9b1342b7a000da05074e4572abe6ce7feb9c is running on https://glitch.cat.family, with no reported issues.

### CODE REVIEW

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
Thanks for the contribution!

The change is welcome even though I would have done without additional helpers and duplicate exceptions (`Mastodon::NotFound` could just be `ActiveRecord::RecordNotFound`).

File: app/lib/search_query_transformer.rb [hunk 1]
```diff
@@ -88,14 +88,16 @@ def initialize(prefix, term)
...
         @filter = :account_id
-        username, domain = term.split('@')
-        account = Account.find_remote(username, domain)
 
-        raise "Account not found: #{term}" unless account
+        username, domain = Account.validate_account_string(term)
+        raise Mastodon::SyntaxError if username.nil?
+
+        account = Account.find_local_or_remote(username, domain)
+        raise Mastodon::NotFound if account.nil?
```
[the remark this answers is not in the document]
Comment (single-right-quote) [reply app/lib/search_query_transformer.rb:96]: I&#x2019;m not familiar with project etiquette here. I have recreated your suggestion locally (and removed code I added that it renders unused), but should I instead commit your suggestion from GitHub, then removed the needless extra code. and then squash?

**Comment by ClearlyClaire** [comment]
Honestly, do as is more convenient for you (maybe avoiding force-pushes though). The PRs get squashed once they get accepted, so it is not necessary for you to do so.

File: app/lib/search_query_transformer.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/lib/search_query_transformer.rb:96]: Honestly, do as is more convenient for you (maybe avoiding force-pushes though). The PRs get squashed once they get accepted, so it is not necessary for you to do so.

[the remark this answers is not in the document]
Comment (single-right-quote) [reply app/lib/search_query_transformer.rb]: Thank you; pushed.

**Review by ClearlyClaire — APPROVED** [review]
I'd rather have an “invalid” prefix either silently fall back to textual comparison, or have a more informative error, but it's already a noticeable improvement over the current situation.

**Comment by single-right-quote** [comment]
I agree. An example where this can be a problem is searching for links, since `https://` is interpreted as a prefix, so either special casing that to fall back (ew) or just falling back in general seems very much worth exploring to me.

**Pull request merged**